ENH: Specific logger for scipy.optimize.shgo

VegardFalmaar opened the pull request:
#### Using the `logging` library

I would like to commend the use of the `logging` library to log messages in SHGO. It seems to me a very good practice which allows the user much flexibilty and useful insight for development and debugging. I have implemented a few changes which enhance the flexibilty on the user's end.


#### Specific SHGO logger

Set up the `scipy.optimize.shgo` solver to log messages to a logger named "scipy.optimize.shgo" instead of the "root" logger. This provides the end user more flexibility in setting up a logging system in their application. When logging to the "root" logger, any changes to the logging level will affect all other loggers in the system as they all inherit settings from "root". Thus the loggers from different libraries, as well as the user's own code, may conflict.

Previously, all the log statements were hidden behind conditional statements using the 'disp' option passed as argument to the `shgo` function. Thus to see the log statements the user had to both set this option to `True` as well as set the logging level of the "root" logger. I have removed this conditional such that setting the logging level of the "scipy.optimize.shgo" logger is sufficient to configure the log messages. I have edited the docstring to illustrate the use.
Removing these conditional statements incur a performance cost which is remedied by a change in the formatting of the log messages.


#### Formatting of log messages

Change the log messages to use lazy formatting rather than f-strings. With the new formatting log messages will not be created unless they are actually going to be logged. This gives a performance improvement for messages that ultimately are not logged. (Running a simple test of the time required to log the elements of a NumPy array with five elements showed that formatting with f-strings takes 45-50 times as long as lazy formatting on my desktop.)

The combined changes in this PR result in practically indistinguishable execution times on my desktop for the benchmarking and test suites (the change is within 1% for the benchmarking suite; I see greater variation between individual runs caused by other processes on the computer).


#### Logging in SciPy

This system of logging could well be extended to the rest of the Python code in SciPy, but I could not find discussions of this elsewhere. The loggers are arranged in a namespace such that "scipy.optimize.shgo" is the child of "scipy.optimize" which is the child of "scipy". Configurations made for a logger automatically affect its children. Thus loggers for different parts of the code base may be set up in the "scipy" namespace and be configured all together, in groups, or individually depending on the user's desires.

VegardFalmaar committed:
ENH: Set up specific logger for scipy.optimize.shgo.

Set the modules directly associated with SHGO to log to the specific
logger named "scipy.optimize.shgo" instead of the "root" logger. This
allows the user to specify the logging level
(and format, output files etc.) for only this routine without affecting
loggers that may be set up in other parts of the user's code. When
logging to the "root" logger, any changes made to the logger will affect
all other loggers as they all inherit from "root".

VegardFalmaar committed:
MAINT: Use lazy logging in scipy.optimize.shgo

Modify logging statements currently using f-strings to use lazy logging
instead. This avoids producing the log statement unless it is actually
going to be logged, giving a significant increase in performance when
the messages are not logged. This eliminates the performance-related
incentive to hide log statements behind conditionals or leave them
commented out.

ilayn commented:
Thanks @VegardFalmaar this is indeed something we really needed. I have been doing a similar thing for SLSQP. So this is good timing. 

One detail that we have to be really careful is how to collate the logs to the same logging mechanism if the users would like to provide their own. Often times this logging works great by itself but say I'm on a remote device and I also have other logs generating stuff and I want to use the same logger then it is an unnecessary pain to find all the relevant loggers and deal with them individually. 

Still having proper logging is much better than just `print`ing but while we are at it we should not paint ourselves into another corner.

VegardFalmaar commented:
Thank you for the quick reply, @ilayn!

I am not quite sure I understand what you mean by your last two paragraphs, so please do correct me if I am mistaken in the following. One always has the opportunity to use the same logging mechanism for all loggers by configuring the "root" logger since it is the parent of all other loggers. If a user configures the "root" logger to use a specific handler then logs from all loggers would be written to this. Alternatively they may cherry pick certain loggers if they need to, and add the same handler to them to collate their messages.

In any case, I am planning to continue working on the SHGO solver. Is there anything regarding log statements that I should do differently as I move forward?

ilayn commented:
> One always has the opportunity to use the same logging mechanism for all loggers by configuring the "root" logger since it is the parent of all other loggers.

The common practice in the industry is to not use root logger. Because you don't know what other loggers are running in other jobs hence you want to provide the algorithm your logger to say "here is my logger use this one", or if not use a standard one and not create a specific one because then you need to intercept this logger if running under other packages and becomes a big mess.

rkern commented on scipy/optimize/_shgo.py:
```diff
@@ -200,7 +200,49 @@ def shgo(
-            Set to True to print convergence messages.
+            Set to True to print convergence messages. This setting applies
+            mainly to compiled subroutines used during optimization. The
+            native Python code is mainly set up to log messages using the
```
This is too much information for a docstring, and all of the logging discussion is irrelevant to `disp`. Elsewhere in the docstring (probably `Notes`), just mention that the Python parts use the standard `logging` framework and log to `scipy.optimize.shgo`. Do not include demos of how to set up logging.

VegardFalmaar replied:
Thank you, I agree.

rkern commented on scipy/optimize/_shgo_lib/_vertex.py:
```diff
@@ -237,9 +238,9 @@ def __getitem__(self, x, nn=None):
-            # is commented out
+            logging.getLogger('scipy.optimize.shgo').info(
+                "New generated vertex at x = %s", x
+            )
```
What is the performance difference? Is the old deleted comment irrelevant with the way that you set this up?

Generally, we'd do `logger = logging.getLogger('scipy.optimize.shgo')` up at the top-level of the module, not create one at each logging statement.

VegardFalmaar replied:
Yes, I would say the old comment is irrelevant. I did some trial runs of the benchmarking suite during development, and the relative difference in execution time with and without this logging statement was less than 1% which is less than the typical variation between individual runs of the same code on my machine.

tupui commented:
Good idea @VegardFalmaar. In general it would be helpful to uniformize logging in the module. As its logs, I suppose there is not much harm experimenting with a given function first and then bringing the solution to other functions.

One thing to consider as well is the level of various messages. I suppose we could move most logs to debug.

VegardFalmaar committed:
MAINT: Define optimize.shgo logger at the top of module

VegardFalmaar committed:
DOC: Shorten disp/logging info in docstring of optimize.shgo

ilayn commented:
> By default, that configuration on the root logger will cascade down to all of the named Loggers (unless if specifically overridden).

This is what you do only if you are in control of the script that you are running and if you are running scipy.optimize code yourself. But anyways, no need to turn this into a logger discussion. I don't have the capacity to continue it further. We can always modify things if they turn out to be problematic. Having loggers instead of printing things is the nice thing already.

VegardFalmaar committed:
MAINT: Move optimize.shgo logs from info to debug

Move most of the logging statements in optimize.shgo from info to debug.

Add clear separation between the iterations in the logs.

Remove duplicate logging statements.

Clarify some conditional statements, and add TODOs for possible future
refactoring.

VegardFalmaar commented:
Thank you for the feedback @tupui! I have moved almost all logging statements to debug in the latest commit. I have also tried to clean up the logs a little bit, e.g. by removing some duplicate logging statements.

lucascolley commented:
Hi @VegardFalmaar , Mypy isn't too happy, here's an excerpt from the CI log:

```
💻  mypy.api.run --config-file /home/runner/work/scipy/scipy/mypy.ini scipy
scipy/optimize/_shgo.py:665: note: By default the bodies of untyped functions are not checked, consider using --check-untyped-defs  [annotation-unchecked]
scipy/optimize/_shgo.py:666: note: By default the bodies of untyped functions are not checked, consider using --check-untyped-defs  [annotation-unchecked]
scipy/optimize/_shgo.py:667: note: By default the bodies of untyped functions are not checked, consider using --check-untyped-defs  [annotation-unchecked]
scipy/optimize/_shgo.py:669: note: By default the bodies of untyped functions are not checked, consider using --check-untyped-defs  [annotation-unchecked]
scipy/optimize/_shgo.py:670: note: By default the bodies of untyped functions are not checked, consider using --check-untyped-defs  [annotation-unchecked]
scipy/optimize/_shgo.py:793: note: By default the bodies of untyped functions are not checked, consider using --check-untyped-defs  [annotation-unchecked]
scipy/optimize/_shgo.py:1096: error: Incompatible types in assignment (expression has type "ndarray[Any, Any]", variable has type "List[ndarray[Any, dtype[signedinteger[_64Bit]]]]")  [assignment]
scipy/optimize/_shgo.py:1098: error: Too many arguments for "tuple"  [call-arg]
scipy/optimize/_shgo.py:1099: error: Need type annotation for "points" (hint: "points: Dict[<type>, <type>] = ...")  [var-annotated]
scipy/optimize/_shgo.py:1106: error: "Tuple[Any, ...]" has no attribute "points"  [attr-defined]
scipy/optimize/_shgo.py:1106: error: "Tuple[Any, ...]" has no attribute "simplices"  [attr-defined]
scipy/optimize/_shgo.py:1158: error: Incompatible types in assignment (expression has type "ndarray[Any, dtype[Any]]", variable has type "List[Any]")  [assignment]
scipy/optimize/_shgo.py:1159: error: Incompatible types in assignment (expression has type "ndarray[Any, dtype[Any]]", variable has type "List[Any]")  [assignment]
scipy/optimize/_shgo.py:1231: error: Incompatible types in assignment (expression has type "ndarray[Any, dtype[Any]]", variable has type "List[Any]")  [assignment]
```

Maybe try pushing a fix for these errors (you can check locally with `python dev.py mypy`) then we can see how the rest of the CI checks are looking.
